cli/crates/backend/src/project.rs

```diff
@@ -65,3 +125,129 @@ fn to_project_path(name: Option<&str>) -> Result<PathBuf, BackendError> {
         None => Ok(current_dir),
     }
 }
+
+enum TemplateInfo<'a> {
+    Grafbase {
+        path: &'a str,
+    },
+    External {
+        repo: &'a str,
+        path: Option<&'a str>,
+        branch: &'a str,
+    },
+}
+
+#[tokio::main]
+async fn download_github_template(template_info: &TemplateInfo<'_>) -> Result<(), BackendError> {
+    let (repo, path, branch) = match template_info {
+        TemplateInfo::Grafbase { path } => ("grafbase/grafbase", Some(PathBuf::from("templates").join(path)), "main"),
+        TemplateInfo::External { repo, path, branch } => (*repo, path.map(PathBuf::from), *branch),
+    };
+
+    let (_, repo_without_org) = repo.split_once('/').expect("must have a slash");
+
+    let extraction_dir = PathBuf::from_str(&format!("{repo_without_org}-{branch}")).expect("must succeed");
+
+    let extraction_result = stream_github_archive(repo, path, branch, extraction_dir.as_path()).await;
+
+    if extraction_dir.exists() {
+        tokio::fs::remove_dir_all(extraction_dir)
+            .await
+            .map_err(BackendError::CleanExtractedFiles)?;
+    }
+
+    extraction_result
+}
+
+async fn stream_github_archive<'a>(
+    repo: &'a str,
```

Comment:
OPT: I wouldn't pass both `path` and `extraction_dir`, they're defining the path within the archive as far as I understand. This path is defined in two steps:
```
    let extraction_dir = PathBuf::from_str(&format!("{repo_without_org}-{branch}")).expect("must succeed");
```
and later
```
    let mut template_path = PathBuf::new();

    template_path.push(extraction_dir);

    if let Some(path) = path {
        template_path.push(path);
    }

    template_path.push("grafbase");
```
I would have combined both to define it in a single place.

Reply:
Done